### PR TITLE
Improve block inserter keyboard navigation

### PR DESCRIPTION
**Pull request by diegohaz**
## The problem

Currently, the block inserter (the one you open by clicking on the `Add block` button at the header toolbar) is visually represented as a two-dimensional grid. To reduce the number of tab stops, each category is implemented as a listbox, so keyboard and screen reader users can use arrow keys to move through the options.

The problem is that, for sighted keyboard users — who can see that there are items above and below the currently focused one —, moving to the right, and not down, by pressing the down arrow key, can be problematic.

Moving through the current block inserter with down arrow key

## This solution

This PR implements two-dimensional keyboard navigation on the block inserter based on https://github.com/WordPress/gutenberg/issues/24975#issuecomment-722481740 and https://github.com/WordPress/gutenberg/issues/21080#issuecomment-649231549 (please, refer to these issues for more context on the implementation and alternatives).

Navigating through the block inserter on this branch with tab and vertical and horizontal arrow keys

## How to test

Try the block inserter keyboard navigation (using arrow keys) on different browsers and screen readers.

### CODE REVIEW

File: packages/block-editor/src/components/block-types-list/style.scss [hunk 1]
```diff
@@ -1,5 +1,4 @@
-.block-editor-block-types-list {
-	list-style: none;
+.block-editor-block-types-list > * {
```
Comment (kevin940726) [inline packages/block-editor/src/components/block-types-list/style.scss:1]: Should we use `*` here or perhaps `[role="presentation"]` instead?

File: packages/block-editor/src/components/block-types-list/index.js [hunk 2]
```diff
@@ -21,49 +24,37 @@ function BlockTypesList( {
...
-						onMouseLeave={ () => onHover( null ) }
-						onBlur={ () => onHover( null ) }
-						isDisabled={ item.isDisabled }
-						title={ item.title }
-						composite={ composite }
-					/>
-				);
-			} ) }
+			{ chunk( items, 3 ).map( ( row, i ) => (
+				<InserterListboxRow key={ i } role="presentation">
```
Comment (kevin940726) [inline packages/block-editor/src/components/block-types-list/index.js:33]: I believe the `role` prop has already been assigned in the `<InserterListboxRow>` component, we don't need to assign it again?

File: packages/block-editor/src/components/block-types-list/index.js [hunk 3]
```diff
@@ -21,49 +24,37 @@ function BlockTypesList( {
...
+								onSelect( item );
+								onHover( null );
+							} }
+							onFocus={ () => onHover( item ) }
+							onMouseEnter={ () => onHover( item ) }
+							onMouseLeave={ () => onHover( null ) }
+							onBlur={ () => onHover( null ) }
+							isDisabled={ item.isDisabled }
+							title={ item.title }
+							isFirst={ i + j === 0 }
```
Comment (kevin940726) [inline packages/block-editor/src/components/block-types-list/index.js:51]: Nitpick: I would prefer just using `i === 0 && j === 0` but I think this is also readable IMO.

**Review by kevin940726 — COMMENTED** [review]
I love this! Haven't tried it locally though, just some minor nitpicking.

**Comment by tellthemachines** [comment]
Tested on macOS with VoiceOver/Safari and Windows 10 with NVDA/Firefox. The keyboard-only interaction is great, it's much easier to get to a specific block. 

With the screen readers, things get a bit confusing: each section still has a role of `listbox`, but the keyboard interaction is no longer what would be expected in a listbox. Arrow up /down no longer takes us to the previous/next element on the list, and can even take us out of the list altogether, and into the next list. 

If we implement this change, we should replace the `listbox` role with something that makes more sense with this keyboard pattern. Perhaps an [ARIA grid](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Grid_Role)? Not sure how good screen reader support for that pattern is though. Would be good to get some further a11y feedback on this.

**Comment by diegohaz** [comment]
@kevin940726 Thanks! I've made the changes.

@tellthemachines Thanks a lot for testing it. Regarding your comments, there's more context on https://github.com/WordPress/gutenberg/issues/24975#issuecomment-722481740.

My initial idea was to leverage the grid pattern and use `role="grid"`. But there's been a lot of controversial conversations around this pattern, most notably this blog post: [ARIA Grid As an Anti-Pattern](https://adrianroselli.com/2020/07/aria-grid-as-an-anti-pattern.html). After testing it, I could confirm that row/column announcements are pretty much useless and sometimes confusing as there's no relationship between them other than their visual arrangement.

I understand that using vertical arrow keys may be confusing. JAWS announces the listbox orientation as "horizontal", which gives a hint that horizontal arrow keys should be used instead. All screen readers will announce the number of the item in the set. That is, moving down from item 1 will announce item 4, which is another hint. But we may have to add aria descriptions to make this interaction more clear.

The multiple listboxes is a technique I found to mimic ARIA 1.2 listbox groups, which VoiceOver still doesn't support.

**Review by gziolo — COMMENTED** [review]
Nice work, looking forward to get it merged 👏

It looks like `reakit` is getting more widely used in Gutenberg which I can only applaud 🎉 I see it's being consumed in 4 packages as of this PR. Separately from this effort we should discuss whether we already need to introduce strategy to avoid code duplication in bundled code. It might be non-issue if only low-level APIs are used knowing that Reakit does great job optimizing for dead-code elimination. However, I can see the scenario when code that depends on `Popper.js` gets used in a few packages and it suddenly becomes more concerning. I know that @ItsJonQ is thinking about extracting a package with a11y tools so that efforts should include what I have just shared. There isn't anything actionable for this PR as far as I'm concerned.

**Comment by tellthemachines** [comment]
> I understand that using vertical arrow keys may be confusing. JAWS announces the listbox orientation as "horizontal", which gives a hint that horizontal arrow keys should be used instead. All screen readers will announce the number of the item in the set. That is, moving down from item 1 will announce item 4, which is another hint. But we may have to add aria descriptions to make this interaction more clear.

VoiceOver and NVDA don't announce the listbox orientation though. Pressing Down arrow and going from "one of ten" to "four of ten", or worse, from "seven of ten" to "one of seven" will just seem like a broken UI for the majority of screen reader users. It will require a fairly verbose description to give a clue as to what's going on here. 

Or perhaps we could resort to an accordion pattern, with all the panels expanded by default, and simple tabbable lists inside each panel, with optional arrow key navigation for keyboard-only users. The grid arrow navigation is not going to be much use to screen reader users anyway, unless they have memorised the exact position of the block they're after. And in that case, it's much easier to use the slash inserter 🙂

**Comment by diegohaz** [comment]
> VoiceOver and NVDA don't announce the listbox orientation though. Pressing Down arrow and going from "one of ten" to "four of ten", or worse, from "seven of ten" to "one of seven" will just seem like a broken UI for the majority of screen reader users. It will require a fairly verbose description to give a clue as to what's going on here.

Yeah! That's why I thought we would have to add an aria description like `Use left and right arrow keys to navigate through blocks`. Slack does something similar with their emoji picker widget, although I don't think their solution works well on VoiceOver due to how they structured their combo box.

The `"seven of ten" to "one of seven"` issue is only a problem on VoiceOver on macOS, and only if you're using arrow keys. With VO keys, it properly announces the group labels when you change between groups. In fact, in NVDA and JAWS, navigating through [this multiple listbox approach](https://block-inserter-keyboard-navigation.vercel.app/2d-multiple-listbox) is pretty much the same as navigating through a [single listbox with multiple ARIA 1.2 groups](https://block-inserter-keyboard-navigation.vercel.app/2d-single-listbox): the total number of items in the set is only announced relatively to the group, and not to the whole listbox.

**Comment by alexstine** [comment]
I've tested the latest changes, here are my findings.

1. When you first select "Add Block", that list is not accessible with the arrow keys. You can only move through that list using Tab.
2. When searching for a block, using the arrow keys no longer works. Tab is the only way to move through that list.
3. I would make the Search label more specific. Maybe "Search for Blocks" to add a bit more context.
4. When using Tab to move through the block categories, it would be nice to hear "category name, instructions, then the first list item" instead of "instructions, category name, and first block."

That's all for now, thanks.

**Comment by diegohaz** [comment]
@alexstine Thanks a lot for the feedback.

> When you first select "Add Block", that list is not accessible with the arrow keys. You can only move through that list using Tab.

That makes sense. Since this PR aims to improve the navigation while in the block list, I think we should address that in a follow-up PR.

> When searching for a block, using the arrow keys no longer works. Tab is the only way to move through that list.

Good catch! Fixed that.

> I would make the Search label more specific. Maybe "Search for Blocks" to add a bit more context.

I made another PR to discuss and address that (#28393).

> When using Tab to move through the block categories, it would be nice to hear "category name, instructions, then the first list item" instead of "instructions, category name, and first block."

The instructions are added through a `polite` live region on focus. It looks like NVDA will announce the live region update before the category and block names. VoiceOver, though, will announce the live region update last, after the category and block names.

I couldn't find a way to announce the instructions in between them, but I [delayed the instructions a little bit](https://github.com/WordPress/gutenberg/pull/26938/commits/c0d52d1371a00370bfbab19f1f569df397be7035) so NVDA announces just like VoiceOver (category name, first block, and instructions), which I think also makes sense.

**Comment by alexstine** [comment]
NVDA still announces instructions, category name, and then the first item. If there's no way around this, I wouldn't call it a blocker by any means.

Everything else seems to be working fine.

Thanks.

**Comment by diegohaz** [comment]
Hi, @tellthemachines, @gziolo, @kevin940726. 👋  Could you do another review on this?

Also cc @talldan

**Comment by ItsJonQ** [comment]
@diegohaz This is wonderful!

I'm excited to see the `state={}` of Reakit in action here 💪 .

My a11y knowledge is not as deep as some of the folks here. It's always a pleasure learning new things about this, especially through something like Reakit.

I did interaction testing with both voiceover on and off. As others have noted, it's a great improvement!

From a code perspective, I think the structure works very well.
I did have one question though. This has more to do with understanding `Composite`.

In the case of the 3-grid chunked layout of Inserter. Is there a difference between chunking items with [`CompositeGroup`](https://reakit.io/docs/composite/#compositegroup) vs setting up [`useCompositeState`](https://reakit.io/docs/composite/#usecompositestate) with `orientation` enabling all arrow key directions?

Thanks <3

File: packages/block-editor/src/components/inserter-list-item/index.js [hunk 4]
```diff
@@ -20,11 +16,12 @@ import {
  * Internal dependencies
  */
 import BlockIcon from '../block-icon';
+import { InserterListboxItem } from '../inserter-listbox';
 import InserterDraggableBlocks from '../inserter-draggable-blocks';
 
 function InserterListItem( {
 	className,
-	composite,
+	isFirst,
```
Comment (gziolo) [inline packages/block-editor/src/components/inserter-list-item/index.js:24]: `InserterListItem` isn't exposed as part of the public API so it's fine to change 👍🏻

**Review by sarayourfriend — APPROVED** [review]
LGTM! Thanks for the walk through earlier, it was super helpful 🚀

**Review by kevin940726 — APPROVED** [review]
LGTM 👍  Thank you for working on this ❤️

**Comment by diegohaz** [comment]
> In the case of the 3-grid chunked layout of Inserter. Is there a difference between chunking items with [`CompositeGroup`](https://reakit.io/docs/composite/#compositegroup) vs setting up [`useCompositeState`](https://reakit.io/docs/composite/#usecompositestate) with `orientation` enabling all arrow key directions?

@ItsJonQ Yes! The orientation will only enable/disable arrow keys, but pressing vertical arrow keys will still move to left/right. `CompositeGroup` creates a row so you can move down/up with vertical arrow keys.

**Pull request merged**